### PR TITLE
Integrate chart-testing as a chart check

### PR DESCRIPTION
**Pull request by isutton**
The plan is to allow the configuration of the `chartTesting` check to be achieved in two different ways:

1. Through the `--set` command line option, which in addition of configuring `chart-verifier` itself, can set check's configuration values.
    ```text
    $ chart-verifier                                                 \
        verify                                                       \
        --enable chartTesting                                        \
        --set chartTesting.buildId=${BUILD_ID}                       \
        --set chartTesting.upgrade=true                              \
        --set chartTesting.skipMissingValues=true                    \
        --set chartTesting.namespace=${NAMESPACE}                    \
        --set chartTesting.releaseLabel="app.kubernetes.io/instance" \
        some-chart.tgz
    ```

2. Through the `--set-values` command line option, where a YAML file can be used to configure the check; below is an example of such a configuration file.
   \
   Chart verifier configuration file `config.yaml`:
    ```yaml
    chartTesting:
        buildId: <BUILD_ID>
        upgrade: true
        skipMissingValues: true
        namespace: <NAMESPACE>
        releaseLabel: "app.kubernetes.io/instance"
    ```
    \
    Once the configuration is persisted, then a chart can be verified:
    ```text
    $ chart-verifier verify --enable chartTesting --set-values config.yaml some-chart.tgz
    ```

The `chartTesting` check will perform the following actions, keeping the semantics provided by `github.com/helm/chart-testing`:

- Install: the chart being verified will be installed in the available OpenShift cluster utilizing the same semantics `client-go` uses to find the current context: a) `--kubeconfig` flag, b) `KUBECONFIG` environment variable and c) `$HOME/.kube/config`.
  
- Test: once a release is installed for the chart being verified, performs the same actions as `helm test` would, which installing all chart resources containing the `"helm.sh/hook": test` annotation.

The check will be considered successful when the chart's installation and tests are all successful.

### CODE REVIEW

**Comment by mmulholla** [comment]
The approach looks good to me.

File: cmd/verify.go [hunk 1]
```diff
@@ -161,21 +169,25 @@ func NewVerifyCmd(config *viper.Viper) *cobra.Command {
 		},
 	}
 
-	cmd.Flags().StringSliceVarP(&opts.ValueFiles, "chart-values", "f", nil, "specify values in a YAML file or a URL (can specify multiple)")
+	settings.AddFlags(cmd.Flags())
+
+	cmd.Flags().StringSliceVarP(&opts.ValueFiles, "chart-values", "F", nil, "specify values in a YAML file or a URL (can specify multiple)")
```
Comment (pedjak) [inline cmd/verify.go:174]: any reason for changing short arg from `f` to `F`?

**Comment by pedjak** [comment]
chartTesting should be enabled by default, right? After passing the check, we should collect at least openshift/k8s version of the cluster, so that this information can be used in PR submission workflow. Hence, `Result` struct should be able to return that or `CheckFunc` should have some way to report that back.

File: cmd/verify.go (hunk 1 above)
Comment (isutton) [reply cmd/verify.go:174]: To make room for now; I understand the defaults are not good so any ideas on how to name the short options are welcome.

**Comment by isutton** [comment]
The following describes the process taken to verify how `helm test` behaves with a chart containing tests, in this case [Bitnami's Wordpress](https://github.com/bitnami/charts/tree/master/bitnami/wordpress) chart. Before proceeding to the test the chart, some tools should be available in the environment, for example `oc` and `helm` (those will be required for the `chart-testing` check in `chart-verifier`).

To download `oc`, one needs to download it from the [OpenShift Client mirror](https://mirror.openshift.com/pub/openshift-v4/x86_64/clients/ocp/) if for some reason the address <https://downloads-openshift-console.apps-crc.testing/amd64/linux/oc.tar> provided by the OpenShift Console is not available.

The downloaded package contains `oc` and a symbolic link to `kubectl`, and can be unpacked in someone's `$HOME/bin`:

    curl -O https://mirror.openshift.com/pub/openshift-v4/x86_64/clients/ocp/4.7.9/openshift-client-linux-4.7.9.tar.gz
    mkdir -p ~/bin
    tar xfz  openshift-client-linux-4.7.9.tar.gz -C ~/bin
    ls -l ~/bin

    total 142084
    -rw-r--r-- 1 isuttonl isuttonl      954 Apr 25 10:16 README.md
    -rwxr-xr-x 2 isuttonl isuttonl 72733320 Apr 25 10:16 kubectl
    -rwxr-xr-x 2 isuttonl isuttonl 72733320 Apr 25 10:16 oc

To make Helm available as well, the following should be performed (please notice that `helm` stores its binary inside an intermediate directory (`linux-amd64` in this example):

    curl -O https://get.helm.sh/helm-v3.5.4-linux-amd64.tar.gz
    mkdir -p ~/bin
    tar xfz helm-v3.5.4-linux-amd64.tar.gz -C ~/bin
    mv ~/bin/linux-amd64/helm ~/bin
    rm -fr ~/bin/linux-amd64/
    ls -l ~/bin

    total 181252
    -rw-r--r-- 1 isuttonl isuttonl      954 Apr 25 10:16 README.md
    -rwxr-xr-x 1 isuttonl isuttonl 39903232 Apr 14 23:16 helm
    -rwxr-xr-x 2 isuttonl isuttonl 72733320 Apr 25 10:16 kubectl
    -rwxr-xr-x 2 isuttonl isuttonl 72733320 Apr 25 10:16 oc

To install the `nginx` chart provided by Bitnami, which contains the `ci` directory containing the to be tests (cloned in `~/src/charts`):

    helm install --dependency-update wordpress wordpress

    NAME: wordpress
    LAST DEPLOYED: Tue May  4 12:38:03 2021
    NAMESPACE: default
    STATUS: deployed
    REVISION: 1
    NOTES:
    ** Please be patient while the chart is being deployed **
    
    Your WordPress site can be accessed through the following DNS name from within your cluster:
    
        wordpress.default.svc.cluster.local (port 80)
    
    To access your WordPress site from outside the cluster follow the steps below:
    
    1. Get the WordPress URL by running these commands:
    
      NOTE: It may take a few minutes for the LoadBalancer IP to be available.
            Watch the status with: 'kubectl get svc --namespace default -w wordpress'
    
       export SERVICE_IP=$(kubectl get svc --namespace default wordpress --template "{{ range (index .status.loadBalancer.ingress 0) }}{{.}}{{ end }}")
       echo "WordPress URL: http://$SERVICE_IP/"
       echo "WordPress Admin URL: http://$SERVICE_IP/admin"
    
    2. Open a browser and access WordPress using the obtained URL.
    
    3. Login with the following credentials below to see your blog:
    
      echo Username: user
      echo Password: $(kubectl get secret --namespace default wordpress -o jsonpath="{.data.wordpress-password}" | base64 --decode)

The deployment can take a couple of minutes finish, so we need to wait until the deployment is ready:

    oc get all,pvc

    NAME                             READY   STATUS    RESTARTS   AGE
    pod/wordpress-7f97d8bffb-lrdjx   1/1     Running   0          115s
    pod/wordpress-mariadb-0          1/1     Running   0          115s
    
    NAME                        TYPE           CLUSTER-IP     EXTERNAL-IP                            PORT(S)                      AGE
    service/kubernetes          ClusterIP      10.217.4.1     <none>                                 443/TCP                      26d
    service/openshift           ExternalName   <none>         kubernetes.default.svc.cluster.local   <none>                       26d
    service/wordpress           LoadBalancer   10.217.4.212   <pending>                              80:31882/TCP,443:32558/TCP   115s
    service/wordpress-mariadb   ClusterIP      10.217.5.37    <none>                                 3306/TCP                     115s
    
    NAME                        READY   UP-TO-DATE   AVAILABLE   AGE
    deployment.apps/wordpress   1/1     1            1           115s
    
    NAME                                   DESIRED   CURRENT   READY   AGE
    replicaset.apps/wordpress-7f97d8bffb   1         1         1       115s
    
    NAME                                 READY   AGE
    statefulset.apps/wordpress-mariadb   1/1     115s
    
    NAME                                             STATUS   VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS   AGE
    persistentvolumeclaim/data-wordpress-mariadb-0   Bound    pv0022   100Gi      RWO,ROX,RWX                   115s
    persistentvolumeclaim/wordpress                  Bound    pv0018   100Gi      RWO,ROX,RWX                   115s

Once the deployment is ready, the chart's test suits can be executed by running `helm test`:

    helm test wordpress

    NAME: wordpress
    LAST DEPLOYED: Tue May  4 12:38:03 2021
    NAMESPACE: default
    STATUS: deployed
    REVISION: 1
    TEST SUITE:     wordpress-credentials-test
    Last Started:   Tue May  4 12:40:36 2021
    Last Completed: Tue May  4 12:40:40 2021
    Phase:          Succeeded
    NOTES:
    ** Please be patient while the chart is being deployed **
    
    Your WordPress site can be accessed through the following DNS name from within your cluster:
    
        wordpress.default.svc.cluster.local (port 80)
    
    To access your WordPress site from outside the cluster follow the steps below:
    
    1. Get the WordPress URL by running these commands:
    
      NOTE: It may take a few minutes for the LoadBalancer IP to be available.
            Watch the status with: 'kubectl get svc --namespace default -w wordpress'
    
       export SERVICE_IP=$(kubectl get svc --namespace default wordpress --template "{{ range (index .status.loadBalancer.ingress 0) }}{{.}}{{ end }}")
       echo "WordPress URL: http://$SERVICE_IP/"
       echo "WordPress Admin URL: http://$SERVICE_IP/admin"
    
    2. Open a browser and access WordPress using the obtained URL.
    
    3. Login with the following credentials below to see your blog:
    
      echo Username: user
      echo Password: $(kubectl get secret --namespace default wordpress -o jsonpath="{.data.wordpress-password}" | base64 --decode)

This is the baseline of the integration this PR implements, and this is how `chart-testing` as a check should behave.

File: pkg/tool/oc.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
+package tool
+
+import (
+	"fmt"
+
+	"github.com/helm/chart-testing/v3/pkg/exec"
+	"gopkg.in/yaml.v3"
+)
+
+type Oc struct {
```
Comment (pedjak) [inline pkg/tool/oc.go:10]: does it need to be public?

**Review by pedjak — APPROVED** [review]

**Comment by pedjak** [comment]
please squash before merge.

**Comment by isutton** [comment]
> please squash before merge.
> 

Can't we rely on GitHub's "Squash & Merge" functionality instead?

**Comment by isutton** [comment]
> chartTesting should be enabled by default, right? After passing the
> check, we should collect at least openshift/k8s version of the
> cluster, so that this information can be used in PR submission
> workflow. Hence, `Result` struct should be able to return that or
> `CheckFunc` should have some way to report that back.
> 

Yes, it should be enabled by default (see 564fea9). 

There's a caveat though: since the `CertifiedOpenShiftVersions` key
was added in the `ToolMetadata` due to adjustment with the schema in a
previous PR, the version is stated as `N/A` if no suitable OpenShift
configuration is found.

**Pull request merged**